### PR TITLE
Adding HTTPs support to AS2 component

### PR DESCRIPTION
**Pull request by shikhar97gupta**
Trying to add SSLContext option to AS2Configuration.

I added a new SSLContext field to AS2Configuration class and got "Unable to create mojo: Empty doc for option: sslContext, parent options: <null>" error while compiling. Following the contribution page, I have already added javadoc comment before the setter method (didn't see it before any getter method of other fields so skipped it for getter) and @UriParam for the field declaration.

(I had written code for adding the SSLContext configuration at component and endpoint level, but keeping the code minimal here. Also changed SSLContext to a String. Getting the same error though)

### CODE REVIEW

**Comment by davsclaus** [comment]
Okay its due to the speciall name, you can do

```
    @UriParam(description = "The name of the SSLContext", name = "SSLContext")
    private String sslContext;
```

**Comment by shikhar97gupta** [comment]
@davsclaus Are these changes fine? Should I keep separate tests for http and https? In the current changes, I use AS2ClientManagerIT to test http while AS2ServerManagerIT uses https. Both of the tests are passing.

**Comment by davsclaus** [comment]
Yes this is fine IMHO

**Comment by shikhar97gupta** [comment]
I ran 
1) mvn clean install -Pformat,fastinstall
2) mvn clean install -Psourcecheck

But sourcecheck is throwing errors during 2 Kafka ITs:

1.

```[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 22.172 s <<< FAILURE! - in org.apache.camel.component.kafka.integration.KafkaConsumerAuthIT 
[ERROR] org.apache.camel.component.kafka.integration.KafkaConsumerAuthIT.kafkaMessageIsConsumedByCamel  Time elapsed: 10.023 s  <<< FAILURE!
java.lang.AssertionError: mock://result Received message count. Expected: <5> but was: <3>
	at org.apache.camel.component.mock.MockEndpoint.fail(MockEndpoint.java:1860)
	at org.apache.camel.component.mock.MockEndpoint.assertEquals(MockEndpoint.java:1817)
	at org.apache.camel.component.mock.MockEndpoint.doAssertIsSatisfied(MockEndpoint.java:449)
	at org.apache.camel.component.mock.MockEndpoint.assertIsSatisfied(MockEndpoint.java:427)
	at org.apache.camel.component.kafka.integration.KafkaConsumerAuthIT.kafkaMessageIsConsumedByCamel(KafkaConsumerAuthIT.java:134)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.SameThreadTimeoutInvocation.proceed(SameThreadTimeoutInvocation.java:45)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
```
2.

```[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 42.816 s <<< FAILURE! - in org.apache.camel.component.kafka.integration.KafkaConsumerAuthInvalidIT
[ERROR] org.apache.camel.component.kafka.integration.KafkaConsumerAuthInvalidIT.kafkaMessageIsConsumedByCamel  Time elapsed: 30.018 s  <<< ERROR!
java.util.concurrent.TimeoutException: kafkaMessageIsConsumedByCamel() timed out after 30 seconds
	at org.junit.jupiter.engine.extension.TimeoutExceptionFactory.create(TimeoutExceptionFactory.java:29)
	at org.junit.jupiter.engine.extension.SameThreadTimeoutInvocation.proceed(SameThreadTimeoutInvocation.java:58)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
```
I do not how to resolve these 2. I encountered other errors while running sourcecheck but those tests passed on re-run.

**Comment by davsclaus** [comment]
You should NOT test all of camel, only your component itself.

The kafka failure is a known test failure and we have a JIRA about it

**Comment by shikhar97gupta** [comment]
@davsclaus 

Okay. Actually I am not sure what the next steps are from here on (to get the PR merged). Do I just open this up for review now? Do I need to sync and update this PR before it gets merged? Sorry if the question is too trivial, I looked it up on stack overflow but only got more confused.

**Comment by davsclaus** [comment]
If you think the PR is ready then you can mark it ready for review

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2ClientConnection.java [hunk 1]
```diff
@@ -69,8 +73,16 @@ public AS2ClientConnection(String as2Version, String userAgent, String clientFqd
...
-        Socket socket = new Socket(targetHost.getHostName(), targetHost.getPort());
-
+        //Socket socket = new Socket(targetHost.getHostName(), targetHost.getPort());
+	
+	if (sslContext==null) {
+		socket = new Socket(targetHost.getHostName(), targetHost.getPort());
+	} else {
+		SSLSocketFactory factory = sslContext.getSocketFactory();
+		socket = (SSLSocket) factory.createSocket(targetHost.getHostName(), targetHost.getPort());
+		//socket.setEnabledProtocols(new String[]{"TLSv1.3"});
```
Comment (oscerd) [inline components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2ClientConnection.java:83]: Why is it commented?

**Review by oscerd — CHANGES_REQUESTED** [review]
Just one question, it seems to be fine overall

File: components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2ClientConnection.java (hunk 1 above)
Comment (shikhar97gupta) [reply components/camel-as2/camel-as2-api/src/main/java/org/apache/camel/component/as2/api/AS2ClientConnection.java:83]: I was having some issues with setting up KeyStore/ TrustStore and got handshake failed error. So I tried changing allowed protocol to only TLSv1.3. 
Did not need it though, so I commented it at that time and forgot to remove this line.

**Review by oscerd — APPROVED** [review]

**Review by oscerd — CHANGES_REQUESTED** [review]
You need to resolve conflicts and rebase

**Comment by davsclaus** [comment]
@shikhar97gupta main branch has been very active due to camel v4 development so this PR needs to be rebased, or if you have too much trouble you can create a new PR, or if you prefer to work on stable 3.x, then use the camel-3.x branch instead.

**Comment by McIntozh** [comment]
@shikhar97gupta, I was able to implement your changes in the current v4 main branch:
[Compare McIntozh/CAMEL-17946 to main](https://github.com/apache/camel/compare/main...McIntozh:camel:CAMEL-17946)
Unfortunately I found, that your solution only works when the host (like "localhost" in your modified unit test) is trusted by the default HostNameVerifier.
I checked against the mendelson test instance at http://mendelson.de/as2_testserver which uses a self signed cerificate and I only got it to work, applying a NoopHostnameVerifier in the component. I don't think we should hard code a NoopHostnameVerifier here. Maybe the verifier should also be supplied via configuration?

**Comment by davsclaus** [comment]
Another PR has added https support to camel-as2
https://issues.apache.org/jira/browse/CAMEL-17946

**Pull request closed without merging**